Docs: Release process (#20934)
grafana/grafana: RELEASES.md

# Grafana Releases

> 

Reviewer: Add an intro that tells the reader what this document will tell them. Pretend you are explaining it to a brand new employee.

And as a general practice, avoid stacked headings. Either add an intro or remove one of the headings.

## Release schedule

> Release cadence of first beta-release being cut is 6 weeks.

Reviewer: Unclear what "being cut" means.

Also, what happens during those six weeks? @daniellee  told me there was a five-week cadence with one week rest. Is the first week usually planning?

You don't have to write yourself into a strict schedule, just give our imaginary new employee reader an idea of what that six-week release is like so they know what to expect.

## Branch management and versioning strategy

We use [Semantic Versioning](https://semver.org/).

We maintain a separate branch for each minor release, named `v<major>.<minor>.x`, e.g. v6.0.x, v6.5.x. The release branch is created from the master branch just before  releasing the first beta-release of a minor version, e.g. v6.5.0-beta1.

> New features, bug fixes and changes are squashed and merged into the master branch. Bug fixes needed to be included in a patch release are being cherry-picked from the master branch. The master branch is the development branch.

Reviewer: ```suggestion
New features, bug fixes, and changes are squashed and merged into the master branch. Bug fixes required in a patch release are being cherry-picked from the master branch. The master branch is the development branch.
```

## Release management roles

> 

Reviewer: Again, add some intro text. We want to avoid stacked headings.

> ### Release coordinator

Reviewer: This is confusing. The first heading under Release management roles is an actual role title, but the second one does not look like a role title. Please make consistent.

Reviewer: Also, I suggest capitalizing the roles so it is clear throughout that these are roles or titles.

Release Coordinator
Release Responsible

Author: Yeah I know. Tried to come up with a good role name of the one doing the actual release. This was my best alternative :) Release manager feels like something else, right?

Do you have a better suggestion?

Reviewer: Release Executor? That sounds all badass. I will see if anyone on Slack has ideas.

> Responsible for coordinating entire release series of a minor release, meaning all beta- and patch releases of a minor release.

Reviewer: ```suggestion
Person responsible for coordinating the entire release series of a minor release, meaning all beta and patch releases of a minor release.
```

> ### Release responible

Reviewer: ```suggestion
### Release responsible
```

> Assigned by release coordinator or voluntarily requested. Responsible for doing the actual work of an individual release.

Reviewer: ```suggestion
Person assigned by the release coordinator or who volunteers. They are responsible for doing the actual work of an individual release.
```

Reviewer: Is this only one person per release?

> ## Create a new milestone

Reviewer: It looks like you are starting a description of a process.

Create a new milestone
Pre-release tasks
Check milestone status
and so on

I suggest moving the level 2 headings down one level and collecting them under something that describes the "bucket" they are in, like "Release process."

You could also rewrite it as an ordered list, which would save you from having to add content to the empty heading sectins.

> **What:** Create a milestone to keep track of every issue and PR to be included in a release

Reviewer: Put periods at the end of sentences. I would also change the order to Who, What, When.

**Who:** Release coordinator or any maintainer

## Before the release

**When:** At least a day before the target release date
**What:** Keep track of milestone process to be able to decide/suggest date of release
**Who:** Release coordinator

### Check status of milestone

> Every PR that's included in milestone have proper labeling:

Reviewer: ```suggestion
Every PR that's included in a milestone must be properly labeled:
```

> - `type/bug` - if the issue it fixes is labeled as a bug

Reviewer: Begin each definition with a capital letter, end all sentences with periods.

- `cherry-pick needed` - denoting that a PR needs to be cherry-picked to be included in release. Applicable for any release after first beta release, e.g. release branch exists.
- `add to changelog` - if the PR should be added to the changelog.

> Any PR merged that haven't been added to the release milestone?

Reviewer: Why are you asking this sentence here? Unclear.

Do you mean that someone should check to see if there are PRs that have been merged that have not been added to the release milestone?

> ### Decide/suggest date of release

Reviewer: Are we deciding or suggesting? And who is doing it?

> ### Assign release responsible

Reviewer: Who is assigning them? I know it was mentioned up in the roles section that the release coordinator assigns the responsibility, but you should also state here something like:

The release coordinator assigns the release responsible role.

## Prepare the release

> **Preconditions:**

Reviewer: If this is presented in an ordered list and is part of the process, then by virtue of where it is, the previous steps should be assumed to have been done. You can probably remove the Preconditions section.

  - Date of release decided
  - Release responsible assigned

**When:** At least a day before the target release date
**What:** Prepare everything needed to be ready to draft a new release on the target release date

> **Who:** Release coordinator and release responsible(s)

Reviewer: ```suggestion
**Who:** Release coordinator and release responsible
```

Reviewer: "Responsible" does not pluralize well, it also is not a noun. Please rethink using it as a role title. Or at least add "person" or "people" to the end of it.

> ### Update changelog

Reviewer: All these empty headings either need content or need to not be headings.

Also, I recommend using sentences to make clear who is doing what here.

Author: Yeah sorry should have been more clear that I basically just fleshed out the raw structure now. All headings will have content eventually.

> ### First major/minor beta release?

Reviewer: Why are we asking questions? Who is asking the question? What does it mean to me and the process if the answer is one or the other? Unclear.

> ### Any release after first beta release?

Reviewer: See previous comment.

#### Cherry pick commits

## Tag a new release

**Who:** Release responsible(s)
**When:** Target release date

See [Grafana Release Guide](https://).

> ## After the release

Reviewer: Ater the release what? After the release is live in production?
